store.go

```diff
@@ -0,0 +1,128 @@
+package pluginapi
+
+import (
+	"database/sql"
+	"math/rand"
+	"sync"
+	"time"
+
+	// import sql drivers
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
+
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+	"github.com/pkg/errors"
+)
+
+// StoreService exposes the underlying database.
+type StoreService struct {
+	initialized bool
+	api         plugin.API
+	mutex       sync.Mutex
+
+	masterDB  *sql.DB
+	replicaDB *sql.DB
+}
+
+func NewStore(api plugin.API) *StoreService {
+	return &StoreService{api: api}
+}
+
+// Gets the master database handle.
+//
+// Minimum server version: 5.16
+func (s *StoreService) GetMasterDB() (*sql.DB, error) {
+	if err := s.initialize(); err != nil {
+		return nil, err
+	}
+
+	return s.masterDB, nil
+}
+
+// Gets the replica database handle.
+// Returns (nil, nil) if no replica is configured.
+//
+// Minimum server version: 5.16
+func (s *StoreService) GetReplicaDB() (*sql.DB, error) {
+	if err := s.initialize(); err != nil {
+		return nil, err
+	}
+
+	return s.replicaDB, nil
+}
+
+// Close closes any open resources. This method is idempotent.
+func (s *StoreService) Close() error {
+	if !s.initialized {
+		return nil
+	}
+
+	if err := s.masterDB.Close(); err != nil {
+		return err
+	}
+
+	if s.replicaDB != nil {
+		if err := s.replicaDB.Close(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *StoreService) initialize() error {
+	// This check is not atomic; not worth the extra complexity here
+	// considering this method is not called very often.
+	if s.initialized {
+		return nil
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	if !s.initialized {
+		if s.api.GetLicense() == nil {
+			return errors.New("this feature requires a valid enterprise license")
+		}
+
+		config := s.api.GetUnsanitizedConfig()
+
+		// Set up master db
+		db, err := setupConnection(*config.SqlSettings.DataSource, config.SqlSettings)
+		if err != nil {
+			return errors.Wrap(err, "failed to connect to master db")
+		}
+		s.masterDB = db
+
+		// Set up replica db
+		if len(config.SqlSettings.DataSourceReplicas) > 0 {
+			replicaSource := config.SqlSettings.DataSourceReplicas[rand.Intn(len(config.SqlSettings.DataSourceReplicas))]
```

Comment:
Instead of random, should we just always connect to the first one? Or should we copy the server semantics and always connect to every replica?

Reply:
if we do what server does then we'll have to use the roundrobin strategy when anyone calls `GetReplciaDB` or something equivalent. We will also have to reserve more connections from the server. Suggesting to keep things simple. Don't mind connecting to the first.